GDP-2: Regions (#24)
GridTools/gt4py: docs/GDPs/gdp-0002-regions.rst

    # Requires a python function to call stencils on each region
    def ubke(uc, vc, ut, ub, dt5, dt4, grid):​
        domain_y_edge = (grid.ni, 1, grid.nz)
        domain_x_edge = (1, grid.nj, grid.nz)
        main_ub(uc, vc, grid.cosa, grid.rsina, ub, dt5=dt5, ​
                origin=(grid.local_istart, grid.local_jstart, 0),
                domain=(grid.ni, grid.nj, grid.nz))​
        if grid.west_edge:​
            x_edge_ub(ut, ub, dt5=dt5, ​origin=(grid.local_istart, grid.local_jstart, 0), ​domain=domain_x_edge)​
        if grid.south_edge:
            y_edge(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jstart, 0), domain=domain_y_edge)
        if grid.north_edge:
            y_edge_ub(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jend, 0), domain=domain_y_edge)
        if grid.east_edge:
            x_edge_ub(ut, ub, dt5=dt5, origin=(grid.local_iend, grid.local_jstart, 0), domain=domain_x_edge)

The specific feature that we are proposing is the addition of a ``with horizontal()`` context that specifies the horizontal iteration space (over parallel axes) using ``region`` objects.

> The ``region`` object's ``__getitem__`` method uses `splitters` to select a region of the stencil computational domain in the parallel ``I`` and ``J`` axes.

Author: Oh, *splitters* there should instead be *axis offsets*. Thanks for the suggestion!

Using this specification, the example is transformed into:

.. code-block:: python

    istart = I[0] if grid.west_edge else None
    jstart = J[0] if grid.south_edge else None
    iend = I[-1] if grid.east_edge else None
    jend = J[-1] if grid.north_edge else None

    @gtscript.stencil()
    def stencil(uc: Field, vc: Field, cosa: Field, rsina: Field, ut: Field, ub: Field, dt5: float, dt4: float):
        from __externals__ import istart, iend, jstart, jend
        with computation(PARALLEL), interval(...):
            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
            with horizontal(region[istart, :], region[iend, :]):
                ub = dt5 * (ut[0, -1, 0] + ut)
            with horizontal(region[:, jstart], region[:, jend]):
                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])

> Notice that distributed domain decomposition works with this feature - if the external has the value ``None``, then any ``region`` objects using this in its slices ignores it.

Author: Yes, that's it! Note that this is not quite watertight, but good enough for now. We can have a situation of a 3x1 processor layout, then on the middle processor this region will fail to execute: 
```
region[start : stop, :]
```
since `start` and `stop` are both `None`. Do you have a suggestion to get around that?

Detailed Description
--------------------

As introduced above, we propose adding a new ``with horizontal()`` context that specializes the stencil on a region of the horizontal axes bounds using ``region`` objects, which pass information to GT4Py about the horizontal iteration space through the indexing operator, similar to numpy arrays.

Axis Offsets
++++++++++++

Regions computation is specified using `Axis Offsets`, which are defined in GT4Py by subscripting the axes (``I``, ``J``, and ``K``).
These can be indexed, which returns the specific indices within a stencil relative to the compute origin. For example: ``I[0]`` is the first compute point, ``I[1]`` the second, and finally ``I[-1]`` is the last point in the stencil compute domain along the ``I`` axis.

Stencil computation in the horizontal axes behaves differently than in the vertical because statements execute over an index space that may extend beyond the limits defined in the stencil compute domain.
Such ``extents`` cannot be represented by merely subscripting axes, since for example ``I[-1]`` referes to the last compute domain index along the ``I`` axis, not the point before the beginning of it.

> Axis Offsets therefore interally hold another offset, which can be set by adding or subtracting from the subscript.

Reviewer: Maybe clearer?
```suggestion
Axis Offsets therefore internally hold an offset which is added or subtracted from the indexed point in the axis.
```

> For example ``I[0] - 2`` is itself an Axis Offsets that refers to 2 points before the start of the compute domain.

Reviewer: ```suggestion
For example ``I[0] - 2`` is itself an Axis Offset that refers to 2 points before the start of the compute domain in ``I``.
```

> Axis Offsets may be assigned to variables in Python and/or used as externals in GT4Py in order to define ``region``.

Reviewer: Do you mean that an AxisOffset object can be created and manipulated in Python or that AxisOffset indices can be Python variables? It's not 100% clear to me.

Author: Both. They are a proper class with `__add__` `__sub__`, and the r-variants.

If the external variable is set to ``None``, then any regions using that external are ignored.

Region Specification
++++++++++++++++++++

``region`` is a keyword in GT4Py that, when subscripted using slices of axis offsets, defines the restricted computation.
These form the arguments to ``with horizontal()``.

As an example, ``region[I[0], :]`` specifies a restriction to the first compute point on the ``I`` axis, and no restriction in the ``J`` axis.
``I[0]`` is a single point, so when converted to a slice is still the single point.
The ``J`` axis simply has ``:``, which is an unrestricted slice, which GT4Py interprets as an unrestricted axis (behaves normally).

The previous example introduced a key element of regional computation: There must not only be a way of specifying axis offsets outside the compute domain, but slices that extend to infinity in each direction (or alternatively, unrestricted endpoints of axes).

> GTScript interprets an unrestricted start or stop element as extending to infinity (or, unrestricted).

Reviewer: Why is exactly that? Can you provide a short explanation / use case for this requirement?

Author: Sure. Good idea.

        vf(i, j) = &
          (v(i,j) - 0.25*(ua(i-1,j) + ua(i, j))*(cos_sg(i-1,j,3) + cos_sg(i,j,1)))  &
                                *dxc(i,j)*0.5*(sin_sg(i-1,j,3) + sin_sg(i,j,1))
      end do
      if (is == 1) &      ! west-edge
        vf(1, j) = &
          v(1, j)*dxc(1, j)*0.5*(sin_sg(0, j, 3) + sin_sg(1, j, 1))
      if (ie+1 == npx) &  ! east-edge
        vf(npx, j) = &
          v(npx, j)*dxc(npx,j)*0.5*(sin_sg(npx-1, j, 3) + sin_sg(npx, j, 1))
    end do

    do j=js,je+1
      do i=is,ie+1
        divg_d(i,j) = vf(i,j-1) - vf(i,j) + uf(i-1,j) - uf(i,j)
      end do
    end do

    if (gridstruct%sw_corner) &
      divg_d(1,    1) = divg_d(1,    1) - vf(1,    0)
    if (gridstruct%se_corner) &
      divg_d(npx,  1) = divg_d(npx,  1) - vf(npx,  0)
    if (gridstruct%ne_corner) &
      divg_d(npx,npy) = divg_d(npx,npy) + vf(npx,npy)
    if (gridstruct%nw_corner) &
      divg_d(1,  npy) = divg_d(1,  npy) + vf(1,  npy)

    do j=js,je+1
      do i=is,ie+1
        divg_d(i,j) = rarea_c(i,j) * divg_d(i,j)
      end do
    end do

.. code-block:: python

    @gtscript.stencil(...)
    def divergence_corner(...):

>         from __externals__ import istart, iend, jstart, jend

Reviewer: Can you explain the precise meaning and/or expected values for these splitters (istart, iend, ...)?

Author: I added the code from above. It's the same thing. Thanks for catching that!

    with computation(PARALLEL), interval(...):
        with horizontal(region[:I[0], :]):
            u = 0

This will set ``u=0`` in all extended computation points to the left of the compute domain.

Examples of this are shown in the image below.
The blue line shows the compute domain along the ``I`` axis, and two examples of region axis slices are shown in red.

.. image:: _static/gdp-0002/axis_offsets.svg

Execution
+++++++++

Another key feature to remember when using regions is that these should be thought of as specifying specialized computation at points.
These are therefore not guaranteed to execute, except where inside the compute domain.
The statements inside a block with ``region[:I[0]-1, :]`` will only execute where the outputs from that block are necessary to compute something else with an extent.
For example, the following will execute

.. code:: python

    with computation(PARALLEL), interval(...):
        with horizontal(region[I[0]-1, :]):
            field_in = 0.0
        field_out = field_in[-1, 0, 0] + field_in[0, 0, 0]

since the ``field_in`` value at ``I[0]-1`` is being consumed to compute a value of an output field inside the compute domain.

> If this used ``I[0]-2``, the code would be ignored.

Reviewer 2: Is this a safe assumption? What if the user wants that computation to be visible outside the stencil specification? I guess in that case the computation should have been expressed separately. Another way of saying the same is asking if this computation would be executed:

```
with computation(PARALLEL), interval(...):
    with horizontal(region[I[0]-1, :]):
        field = 0.0
```

I can see why you want to optimize the code out, and it's fine with me if this is the semantic we always want

Author: Yes, I would say this is correct. If the user wants it exposed outside the computation then they would have to write the region differently: `region[I[0], :]`, or define an external `ROW=I[0]`, and use it in `region[ROW, :]`.

The idea behind this is that GT4Py stencils are not a complete picture of an application. Talking only about stencils can hide the reality that these are defined on an application grid space, if the operator needs to be specialized on different parts of the grid. Say, for example you want to apply a different finite difference stencil on edges or at corners, which is required for cubed-sphere grid. In this context, the code that goes in regions allows you to define computation with respect to the application grid.

I see two different semantics:
1. Regions define computation no matter what, breaking the natural extent analysis from a single stencil compute domain. I do not yet see through how extent analysis for this method would work. Perhaps every region would spawn a new extent analysis?
2. Define specialized computation at points. If fields indices set in this block are used later, expose this calculation.

The frontend feature currently exposes option (2).

Reviewer 2: Ok, I see. As long as the semantic is clear I have no objections, I wanted to make sure we were aware of things. Thanks!

This information should be collected in a dictionary of field name to numeric tuple, and passed to stencils at execution time.

The call could then look like:

.. code-block:: python

    stencil(uc, vc, cosa, rsina, ut, ub, dt5, dt4,
        app_domain=(grid.ni, grid.nj), app_origins={ub: (1,0)})

Temporary fields
^^^^^^^^^^^^^^^^

The information above is sufficient for field arguments to the stencil, but what about temporaries? Assigning to temporaries in a region should be allowed, so GT4Py needs to generate an offset automatically from the application domain for these fields. The natural way to determine this is to have the `origin` of the temporary field be the origin of the compute domain.

Take for example the double Laplacian example, with a special case before applying the second Laplacian:

.. code-block:: python

    @gtscript.stencil()
    def double_lap(in_f: gtscript.Field[float],
                out_f: gtscript.Field[float]):
        with computation(PARALLEL), interval(...):
            tmp_f = lap(in_f)
            with region(LeftEdge(-1, 0)):

>                 tmp_f = 1.0

Reviewer 3: ```suggestion
                tmp_f = in_f[-1,0,0]
```
Would that be allowed? Why not?

If I replace `tmp_f` by an external field and remove the second lap, will the region be executed or not (i.e. will the external field have 1.0 in the second halo line.

I need to read it more carefully tomorrow, to see if I can find the answer myself.

Author: That should still work, yes. It is accessing the last halo line for `in_f` in that case. It would extend the read extent for `in_f`.

Reviewer 3: Right, now let's trigger a read from one line more (which could just happen by merging another unrelated operation into this stencil). So 3 halo lines which are all consumed. What happens at the second halo line?

Author: Well, that wouldn't be valid anyway (there is no data there in `in_f`). You would have to shrink the computational domain for the stencil. If you didn't specify the domain it would automatically shrink it to `domain=(2,2)`, and the stencil would not error.
